discovery-provider/plugins/notifications/src/tasks/dmNotifications.ts

```diff
@@ -0,0 +1,102 @@
+  const cachedMessageTimestamp = await redis.get(config.lastIndexedMessageRedisKey)
+  if (cachedMessageTimestamp) {
+    minMessageCursor = new Date(Date.parse(cachedMessageTimestamp))
+  }
+  const cachedReactionTimestamp = await redis.get(config.lastIndexedReactionRedisKey)
+  if (cachedReactionTimestamp) {
+    minReactionCursor = new Date(Date.parse(cachedReactionTimestamp))
+  }
+
+  return {
+    maxTimestamp: maxCursor,
+    minMessageTimestamp: minMessageCursor,
+    minReactionTimestamp: minReactionCursor,
+  }
+}
+
+async function getUnreadMessages(discoveryDB: Knex, minTimestamp: Date, maxTimestamp: Date): Promise<DMNotification[]> {
+  return await discoveryDB
+    .select('chat_message.user_id as sender_user_id', 'chat_member.user_id as receiver_user_id', 'chat_message.ciphertext as message', 'chat_message.created_at as timestamp')
+    .from('chat_message')
+    .innerJoin('chat_member', 'chat_message.chat_id', 'chat_member.chat_id')
+    .whereRaw('chat_message.created_at > greatest(chat_member.last_active_at, ?::timestamp)', [minTimestamp.toISOString()])
+    .andWhere('chat_message.created_at', '<=', maxTimestamp.toISOString())
+    .andWhereRaw('chat_message.user_id != chat_member.user_id')
+}
+
+async function getUnreadReactions(discoveryDB: Knex, minTimestamp: Date, maxTimestamp: Date): Promise<DMReactionNotification[]> {
+  return await discoveryDB
+    .select('chat_message_reactions.user_id as sender_user_id', 'chat_message.user_id as receiver_user_id', 'chat_message_reactions.reaction as reaction', 'chat_message.ciphertext as message', 'chat_message_reactions.updated_at as timestamp')
+    .from('chat_message_reactions')
+    .innerJoin('chat_message', 'chat_message.message_id', 'chat_message_reactions.message_id')
+    .joinRaw('join chat_member on chat_member.chat_id = chat_message.chat_id and chat_member.user_id = chat_message.user_id')
+    .whereRaw('chat_message_reactions.updated_at > greatest(chat_member.last_active_at, ?)', [minTimestamp.toISOString()])
+    .andWhere('chat_message_reactions.updated_at', '<=', maxTimestamp.toISOString())
+    .andWhereRaw('chat_message_reactions.user_id != chat_member.user_id')
+}
+
+function setLastIndexedTimestamp(redis: RedisClientType, redisKey: string, maxTimestamp: Date, notifications: MessageNotification[] | MessageReactionNotification[]) {
+  if (notifications.length > 0) {
+    notifications.sort(notificationTimestampComparator)
+    const lastIndexedTimestamp = notifications[notifications.length - 1].notification.timestamp.toISOString()
+    redis.set(redisKey, lastIndexedTimestamp)
+    console.log(`setting last indexed timestamp to ${lastIndexedTimestamp} for key ${redisKey}`)
+  } else {
+    redis.set(redisKey, maxTimestamp.toISOString())
+  }
+}
+
+export async function sendDMNotifications(discoveryDB: Knex, identityDB: Knex) {
+  // Query DN for unread messages and reactions between min and max cursors
+  const redis = await getRedisConnection()
```

Comment:
Does this cost anything to initialize a new redis connection on each call of sendDMNotification - or it is negigible/free to do so?

Reply:
shouldn't be initializing each call; in `redisConnection.ts` it inits once then sets to a global var, then returns the global var for all subsequent `getRedisConnection()` calls